R/survival_qp.R

```diff
@@ -0,0 +1,233 @@
+#' @param lambda Regularization hyperparameter, default 0
+#' @param lowlim Lower limit on weights, default 1
+#' @param uplim Upper limit on weights, default NULL
+#' @param verbose Whether to show messages, default T
+#' @param eps_abs Absolute error tolerance for solver
+#' @param eps_rel Relative error tolerance for solver
+#' @param ... Extra arguments for osqp solver
+#'
+#' @return \itemize{
+#'          \item{weights}{Estimated weights as a length n vector}
+#'          \item{imbalance0}{Imbalance in covariates for control in vector}
+#'          \item(imbalance1){Imbalance in covariates for treated in vector}}
+#' @export
+survival_qp <- function(B_X, trt, times, events, t, lambda = 0, lowlim = 0, uplim = NULL,
+                        verbose = TRUE, eps_abs = 1e-5, eps_rel = 1e-5, ...) {
+  # Convert X to a matrix
+  B_X <- as.matrix(B_X)
+  
+  # Simple data checks
+  check_data_surv(B_X, trt, times, events, t, lambda, lowlim, uplim)
+  
+  n <- nrow(B_X)
+  k <- ncol(B_X)
+  
+  # If no uplim set (default NULL), set uplim to be number of individuals n
+  if (is.null(uplim)) {
+    uplim <- n
+  }
+  
+  # Calculate mean basis vector from basis mx on covariates 
+  Bbar_X <- colMeans(B_X)
+  
+  # Setup the components of the QP and solve
+  if(verbose) message("Creating linear term vector...")
+  q <- create_q_vector_surv(n, trt, B_X, Bbar_X)
+  
+  if(verbose) message("Creating quadratic term matrix...")
+  P <- create_P_matrix_surv(n, trt, B_X)
+  
+  # Add regularization
+  I0 <- create_I0_matrix_surv(n)
+  P <- P + lambda * I0
+  
+  if(verbose) message("Creating constraint matrix...")
+  constraints <- create_constraints(n, trt, times, events, t, lowlim, uplim, verbose)
+  
+  settings <- do.call(osqp::osqpSettings,
+                      c(list(verbose = verbose,
+                             eps_rel = eps_rel,
+                             eps_abs = eps_abs),
+                        list(...)))
+  
+  solution <- osqp::solve_osqp(P, q, constraints$A,
+                               constraints$l, constraints$u,
+                               settings)
+  
+  weights <- solution$x
+
+  # Return vectors of imbalances for treated and control 
+  noncens_t <- ((times >= t) & (events == TRUE)) | (events == FALSE)
+  B0_X <- B_X*(trt == 0)
+  B1_X <- B_X*(trt == 1)
+  
+  imbalance0 <- (t(B0_X) %*% Matrix(weights*noncens_t, n, 1)) - Matrix(Bbar_X, k, 1)
+  imbalance1 <- (t(B1_X) %*% Matrix(weights*noncens_t, n, 1)) - Matrix(Bbar_X, k, 1)
+
+  return(list(weights = weights,
+              imbalance0 = imbalance0[,1],
+              imbalance1 = imbalance1[,1]))
+}
+
+#' Create the q vector for an QP that solves min_x 0.5 * x'Px + q'x
+#' @param B_X n x k basis matrix of covariates 
+#' @param Bbar_X Vector of population means to re-weight to
+#'
+#' @return q vector
+create_q_vector_surv <- function(n, trt, B_X, Bbar_X) {
+  B0_X <- B_X*(trt == 0)
+  B1_X <- B_X*(trt == 1)
+
+  q <- -(2/n) * Bbar_X %*% t(B0_X + B1_X) 
```

Comment:
`osqp` is assuming that you're passing in an optimization problem as $\frac{1}{2}x'Px + q'x$, so this shouldn't have the multiplication by 2, i.e.
```
q <- -(1/n) * Bbar_X %*% t(B0_X + B1_X)
```